### PR TITLE
prom: fixes dynamic registration of views bug

### PR DESCRIPTION
**Pull request by rdallman**
prometheus's default registry appears to block registration of any other
metrics based on there being >= 1 duplicate in the set of metrics returned
from Describe to register. While Describe's API seems to indicate that a
collector should return a superset of metrics, and while this makes perfect
sense with the Unregister logic, it does not seem to play nicely with us
running register on the same collector repeatedly with different metrics, nor
with calling unregister and then attempting to register everything once again.
that is to say, prometheus client seems to explicitly disallow the thing we're
seemingly trying to do. upstream: prometheus/client_golang#47

this patch will:
* no longer call Unregister and then Register, as this is a case handled by
prometheus: https://github.com/prometheus/client_golang/blob/fcc130e101e76c5d303513d0e28f4b6d732845c7/prometheus/registry.go#L112-L117
* only return metrics that are not yet registered to prometheus registry

since ExportView appears to be called in a single thread, I am not sure
the locks are required, but have abided here. that also ruled out this
being a race of any kind.

It's also a perfectly reasonable explanation that I've not properly registered
views. fwiw am doing `v.Subscribe()` for every view in init() atm, and the
intent appears to be to be able to toggle these on/off during runtime.

I was able to generate these errors consistently on the every sweep of
metrics that would be done:

```
cannot register the collector: descriptor Desc{fqName: "redacted", help:
"redacted", constLabels: {}, variableLabels: []} already exists with the same
fully-qualified name and const label values
```

to be clear, this is not due to my attempting to register 2 views / metrics
under the same name, but from registerViews repeatedly registering all metrics
to prometheus registry (confirmed) which will [trivially] contain duplicates.

I apologize that I do not have a reduced reproduction. I do however have a
patch in an OSS project that will reliably reproduce this, though I hate
pointing at that as it'll take some orienting to run.

perhaps these were simply an annoyance and to be ignored, in which case we can
just remove the `onError` bits, I am seeing all metrics in the prometheus
endpoint even though I get this error spat out a lot (and, it's not clear that
this should work this way after reading the prometheus register code). there
may be other solutions :)

thanks for this library, your efforts are appreciated, am looking forward to
using it!

### CODE REVIEW

File: exporter/prometheus/prometheus.go [hunk 1]
```diff
@@ -163,6 +162,10 @@ type collector struct {
 	registeredViewsMu sync.Mutex
 	// registeredViews maps a view to a prometheus desc.
 	registeredViews map[string]*prometheus.Desc
+
+	// unregisteredViews is the set of views that are
```
Comment (rakyll) [inline exporter/prometheus/prometheus.go]: s/unregisteredViews/newViews

Unregistered sounds like they have been registered and unregistered at some point.

File: exporter/prometheus/prometheus.go [hunk 2]
```diff
@@ -163,6 +162,10 @@ type collector struct {
...
+	// queued to be registered with prometheus.
```
Comment (rakyll) [inline exporter/prometheus/prometheus.go]: s/with prometheus/with Prometheus

File: exporter/prometheus/prometheus.go [hunk 3]
```diff
@@ -174,11 +177,19 @@ func (c *collector) addViewData(vd *view.Data) {
 	c.mu.Unlock()
 }
 
+// NOTE that this will not work properly if calling
```
Comment (rakyll) [inline exporter/prometheus/prometheus.go]: Can you move this comment inside the function block?

File: exporter/prometheus/prometheus.go [hunk 4]
```diff
@@ -104,7 +104,7 @@ func (c *collector) registerViews(views ...*view.View) {
 				nil,
 			)
 			c.registeredViewsMu.Lock()
```
Comment (rakyll) [inline exporter/prometheus/prometheus.go]: Can we call registeredViewsMu to viewsMu and document is guards both fields?

**Comment by rdallman** [comment]
addressed, thank you for the review

**Comment by odeke-em** [comment]
> since ExportView appears to be called in a single thread, I am not sure
the locks are required, but have abided here. that also ruled out this
being a race of any kind.

c.viewData can be read/written concurrently by multiple goroutines during a Collect cycle when a user visits /metrics concurrently while ExportView is invoked internally, hence why we need the lock. One can easily reproduce such races as with test https://github.com/census-instrumentation/opencensus-go/issues/402#issuecomment-362905246 and the bug linked there was a real bug reported by @rakyll

**Pull request merged**